tulip/abstract/discretization.py

```diff
@@ -983,6 +1070,474 @@ def discretize(
         disc_params=param
     )
 
+def _discretize_dual(
+    part, ssys, N=10, min_cell_volume=0.1,
+    closed_loop=True, conservative=False,
+    max_num_poly=5, use_all_horizon=False,
+    trans_length=1, remove_trans=False,
+    abs_tol=1e-7,
+    plotit=False, save_img=False, cont_props=None,
+    plot_every=1
+):
+    """Refine the partition and establish transitions
+    based on reachability analysis. Use dual-simulation algorithm.
+
+    Reference
+    =========
+    1. U{[NOTM12]
+    <https://tulip-control.sourceforge.io/doc/bibliography.html#notm12>}
+    2. Wagenmaker, A. J.; Ozay, N.
+       "A Bisimulation-like Algorithm for Abstracting Control Systems."
+       54th Annual Allerton Conference on CCC 2016
+
+    See Also
+    ========
+    L{prop2partition.pwa_partition}, L{prop2partition.part2convex}
+
+    @param part: L{PropPreservingPartition} object
+    @param ssys: L{LtiSysDyn} or L{PwaSysDyn} object
+    @param N: horizon length
+    @param min_cell_volume: the minimum volume of cells in the resulting
+        partition.
+    @param closed_loop: boolean indicating whether the `closed loop`
+        algorithm should be used. default True.
+    @param conservative: if true, force sequence in reachability analysis
+        to stay inside starting cell. If false, safety
+        is ensured by keeping the sequence inside a convexified
+        version of the original proposition preserving cell.
+    @param max_num_poly: maximum number of polytopes in a region to use in
+        reachability analysis.
+    @param use_all_horizon: in closed loop algorithm: if we should look
+        for reachability also in less than N steps.
+    @param trans_length: the number of polytopes allowed to cross in a
+        transition.  a value of 1 checks transitions
+        only between neighbors, a value of 2 checks
+        neighbors of neighbors and so on.
+    @param remove_trans: if True, remove found transitions between
+        non-neighbors.
+    @param abs_tol: maximum volume for an "empty" polytope
+
+    @param plotit: plot partitioning as it evolves
+    @type plotit: boolean,
+        default = False
+
+    @param save_img: save snapshots of partitioning to PDF files,
+        requires plotit=True
+    @type save_img: boolean,
+        default = False
+
+    @param cont_props: continuous propositions to plot
+    @type cont_props: list of C{Polytope}
+    
+    @param simu_type: flag used to choose abstraction algorithm
+        (bisimulation or dual-simulation). 
+    @type simu_type: string, 'bi' or 'dual'
+        default = 'bi'
+    
+    @rtype: L{AbstractPwa}
+    """
+    start_time = os.times()[0]
+
+    orig_ppp = part
+    min_cell_volume = (min_cell_volume /np.finfo(np.double).eps
+        *np.finfo(np.double).eps)
+
```

Comment:
I am in favor of the recommendation by PEP 8 to not leave blank lines whenever possible, to form blocks of lines. I support separating logical blocks of lines, but with an explanatory comment line, instead of a blank one (since vertical space is going to be used, it can be utilized for a comment).